File: backend/tests/integration/vault_secrets_integration_test.go
//go:build integration_vault

package integration

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"github.com/zatekoja/Patientpricediscoverydesign/backend/pkg/secrets"
)

func TestVaultSecretsIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test in short mode")
	}

	addr := firstNonEmpty(os.Getenv("TEST_VAULT_ADDR"), os.Getenv("VAULT_ADDR"))
	token := firstNonEmpty(os.Getenv("TEST_VAULT_TOKEN"), os.Getenv("VAULT_TOKEN"))
	mount := firstNonEmpty(os.Getenv("TEST_VAULT_MOUNT"), "secret")

	if addr == "" || token == "" {
		t.Skip("Vault integration test requires TEST_VAULT_ADDR/TEST_VAULT_TOKEN")
	}

	if !vaultReady(addr) {
		t.Skip("Vault not reachable")
	}

	path := fmt.Sprintf("patient-price-discovery/tests/%d", time.Now().UnixNano())
	data := map[string]string{
		"OPENAI_API_KEY":        "vault-test-openai",
		"GEOLOCATION_API_KEY":   "vault-test-geo",
		"PROVIDER_LLM_API_KEY":  "vault-test-provider",
		"PROVIDER_LLM_ENDPOINT": "https://example.com",
	}

	err := writeVaultSecret(addr, token, mount, path, data)
	require.NoError(t, err)

	prevOpenAI := os.Getenv("OPENAI_API_KEY")
	prevGeo := os.Getenv("GEOLOCATION_API_KEY")
	prevProvider := os.Getenv("PROVIDER_LLM_API_KEY")
	defer restoreEnv("OPENAI_API_KEY", prevOpenAI)
	defer restoreEnv("GEOLOCATION_API_KEY", prevGeo)
	defer restoreEnv("PROVIDER_LLM_API_KEY", prevProvider)

	cfg := secrets.VaultConfig{
		Enabled:   true,
		Addr:      addr,
		Token:     token,
		Mount:     mount,
		Path:      path,
		KVVersion: 2,
		Timeout:   3 * time.Second,
		Overwrite: true,
	}

	result, err := secrets.ApplyVaultSecrets(context.Background(), cfg)
	require.NoError(t, err)
	require.GreaterOrEqual(t, result.Loaded, 3)
	require.Equal(t, "vault-test-openai", os.Getenv("OPENAI_API_KEY"))
	require.Equal(t, "vault-test-geo", os.Getenv("GEOLOCATION_API_KEY"))
	require.Equal(t, "vault-test-provider", os.Getenv("PROVIDER_LLM_API_KEY"))
}

func vaultReady(addr string) bool {
	client := http.Client{Timeout: 2 * time.Second}
	req, err := http.NewRequest(http.MethodGet, strings.TrimRight(addr, "/")+"/v1/sys/health", nil)
	if err != nil {
		return false
	}
	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusTooManyRequests
}

func writeVaultSecret(addr, token, mount, path string, data map[string]string) error {
	payload := map[string]interface{}{
		"data": data,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	base := strings.TrimRight(addr, "/")
	cleanMount := strings.Trim(mount, "/")
	cleanPath := strings.TrimLeft(path, "/")
	url := fmt.Sprintf("%s/v1/%s/data/%s", base, cleanMount, cleanPath)

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Vault-Token", token)

	client := http.Client{Timeout: 3 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("vault write failed: %s", resp.Status)
	}
	return nil
}

func restoreEnv(key, value string) {
	if value == "" {
		_ = os.Unsetenv(key)
	} else {
		_ = os.Setenv(key, value)
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

File: backend/tests/integration/vault_secrets_integration_test.ts
import assert from 'assert';
import crypto from 'crypto';
import { loadVaultSecretsToEnv } from '../../config/vaultSecrets';

async function runTest(name: string, fn: () => void | Promise<void>): Promise<void> {
  try {
    await fn();
    console.log(`✓ ${name}`);
  } catch (err) {
    console.error(`✗ ${name}`);
    console.error(err);
    process.exitCode = 1;
  }
}

async function main() {
  const addr = process.env.TEST_VAULT_ADDR || process.env.VAULT_ADDR || '';
  const token = process.env.TEST_VAULT_TOKEN || process.env.VAULT_TOKEN || '';
  const mount = process.env.TEST_VAULT_MOUNT || 'secret';

  if (!addr || !token) {
    console.log('Skipping integration test: TEST_VAULT_ADDR/TEST_VAULT_TOKEN not set');
    return;
  }

  const ready = await vaultReady(addr);
  if (!ready) {
    console.log('Skipping integration test: Vault not reachable');
    return;
  }

  const path = `patient-price-discovery/tests/${randomId()}`;
  const payload = {
    OPENAI_API_KEY: 'vault-test-openai',
    GEOLOCATION_API_KEY: 'vault-test-geo',
    PROVIDER_LLM_API_KEY: 'vault-test-provider',
  };

  await writeVaultSecret(addr, token, mount, path, payload);

  const prevOpenAI = process.env.OPENAI_API_KEY;
  const prevGeo = process.env.GEOLOCATION_API_KEY;
  const prevProvider = process.env.PROVIDER_LLM_API_KEY;

  await runTest('loadVaultSecretsToEnv loads KV v2 secrets into env', async () => {
    const result = await loadVaultSecretsToEnv({
      enabled: true,
      addr,
      token,
      mount,
      path,
      kvVersion: 2,
      overwrite: true,
      timeoutMs: 3000,
    });

    assert.strictEqual(result.enabled, true);
    assert.strictEqual(result.error, undefined);
    assert.ok(result.loaded >= 3, 'expected secrets to be loaded');
    assert.strictEqual(process.env.OPENAI_API_KEY, payload.OPENAI_API_KEY);
    assert.strictEqual(process.env.GEOLOCATION_API_KEY, payload.GEOLOCATION_API_KEY);
    assert.strictEqual(process.env.PROVIDER_LLM_API_KEY, payload.PROVIDER_LLM_API_KEY);
  });

  restoreEnv('OPENAI_API_KEY', prevOpenAI);
  restoreEnv('GEOLOCATION_API_KEY', prevGeo);
  restoreEnv('PROVIDER_LLM_API_KEY', prevProvider);

  if (process.exitCode) {
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

function restoreEnv(key: string, value?: string) {
  if (value === undefined) {
    delete process.env[key];
  } else {
    process.env[key] = value;
  }
}

function randomId(): string {
  return crypto.randomBytes(6).toString('hex');
}

async function vaultReady(addr: string): Promise<boolean> {
  try {
    const resp = await fetch(`${addr.replace(/\/+$/, '')}/v1/sys/health`);
    return resp.ok || resp.status === 429;
  } catch {
    return false;
  }
}

async function writeVaultSecret(
  addr: string,
  token: string,
  mount: string,
  path: string,
  data: Record<string, string>
): Promise<void> {
  const base = addr.replace(/\/+$/, '');
  const cleanMount = mount.replace(/^\/+|\/+$/g, '');
  const cleanPath = path.replace(/^\/+/, '');
  const url = `${base}/v1/${cleanMount}/data/${cleanPath}`;
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-Vault-Token': token,
    },
    body: JSON.stringify({ data }),
  });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Vault write failed: ${response.status} ${response.statusText} ${text}`);
  }
}
